robots/commenter/main.go

```diff
@@ -236,35 +242,65 @@ func run(c client, query, sort string, asc, random bool, commenter func(meta) (s
 		}
 		issues = dest
 	}
-	for n, i := range issues {
-		if ceiling > 0 && n == ceiling {
-			log.Printf("Stopping at --ceiling=%d of %d results", n, len(issues))
+
+	modifiedIssues := 0
+	for _, i := range issues {
+		if ceiling > 0 && modifiedIssues == ceiling {
+			log.Printf("Stopping at --ceiling=%d of %d results", modifiedIssues, len(issues))
 			break
 		}
 		log.Printf("Matched %s (%s)", i.HTMLURL, i.Title)
-		org, repo, number, err := parseHTMLURL(i.HTMLURL)
-		if err != nil {
-			msg := fmt.Sprintf("Failed to parse %s: %v", i.HTMLURL, err)
-			log.Print(msg)
-			problems = append(problems, msg)
-		}
-		comment, err := commenter(meta{Number: number, Org: org, Repo: repo, Issue: i})
-		if err != nil {
-			msg := fmt.Sprintf("Failed to create comment for %s/%s#%d: %v", org, repo, number, err)
-			log.Print(msg)
-			problems = append(problems, msg)
+		if err := createComment(c, commenter, i, spamCeiling, spamMargin); err != nil {
+			log.Print(err.Error())
+			problems = append(problems, err.Error())
 			continue
 		}
-		if err := c.CreateComment(org, repo, number, comment); err != nil {
-			msg := fmt.Sprintf("Failed to apply comment to %s/%s#%d: %v", org, repo, number, err)
-			log.Print(msg)
-			problems = append(problems, msg)
-			continue
-		}
-		log.Printf("Commented on %s", i.HTMLURL)
+		modifiedIssues++
 	}
 	if len(problems) > 0 {
 		return fmt.Errorf("encoutered %d failures: %v", len(problems), problems)
 	}
 	return nil
 }
+
+func createComment(c client, commenter func(meta) (string, error), issue github.Issue,
+	spamCeiling int, spamMargin time.Duration) error {
+	org, repo, number, err := parseHTMLURL(issue.HTMLURL)
+	if err != nil {
+		return fmt.Errorf("Failed to parse %s: %v", issue.HTMLURL, err)
+	}
+	comment, err := commenter(meta{Number: number, Org: org, Repo: repo, Issue: issue})
+	if err != nil {
+		return fmt.Errorf("Failed to create comment for %s/%s#%d: %v", org, repo, number, err)
+	}
+
+	if spamCeiling > 0 {
+		issueComments, err := c.ListIssueComments(org, repo, number)
+		if err != nil {
+			return fmt.Errorf("Failed to get comments for %s/%s#%d: %v", org, repo, number, err)
+		}
+
+		sameCommentsInSequenceCounter := 0
+		for _, issueComment := range issueComments {
+			if issueComment.Body != comment || issueComment.CreatedAt.Before(time.Now().Add(-1*spamMargin)) {
```

Comment:
Aren't the comments sorted timewise? Meaning, could we break when we hit a comment that's before the margin?

nit: `Sub` instead of `Add(-1)`

Reply:
Yes, I expect the comments to be sorted timewise.
I don't understand what do you mean by the 2nd question. The margin was according to your request on https://github.com/kubernetes/test-infra/pull/23483#discussion_r702483588 .
Regarding the nit - `Sub` has a different API and returns a `Duration` instead of `Time`. I saw this approach in other modules as well.
Additionally, I think that the conversions won't make it essentially cleaner. Let me know if you think differently.

Reply:
> Meaning, could we break when we hit a comment that's before the margin?

I meant that if the comments are sorted, we could `break` from the spam detection loop, because we iterated outside of the margin period, we did not hit the spam threshold until now, and we will not hit it because we now read older comments.

Reply:
Yes, you are right.
According to https://docs.github.com/en/rest/reference/issues#comments 
```
By default, Issue Comments are ordered by ascending ID.
```

Actually, I notice the `since` parameter that we should use instead of running this logic on our side.

I'll add on a new commit a new function that would make this functionality simpler under `github.ListIssueCommentsSince`. Let me know what do you think.